packages/core-data/src/types/README.md

```diff
@@ -0,0 +1,223 @@
+
+Example:
+
+```ts
+interface Post< C extends Context > {
+	…
+	modified: ContextualField< string, 'edit' | 'view', C >;
+	password: ContextualField< string, 'edit', C >;
+	…
+}
+
+const post: Post<'edit'> = …
+// post.modified exists as a string
+// post.password exists as a string
+
+const post: Post<'view'> = …
+// post.modified still exists as a string
+// post.password is missing, undefined, because we're not in the `edit` context.
+```
+
+### OmitNevers
+
+Removes all the properties of type never, even the deeply nested ones.
+
+```ts
+type MyType = {
+  foo: string;
+  bar: never;
+  nested: {
+    foo: string;
+    bar: never;
+  }
+}
+const x = {} as OmitNevers<MyType>;
+// x is of type { foo: string; nested: { foo: string; }}
+// The `never` properties were removed entirely
+```
+
+### Updatable
+
+Updatable<EntityRecord> is a type describing Edited Entity Records. They are like
+regular Entity Records, but they have all the local edits applied on top of the REST API data.
+
+This turns certain field from an object into a string.
+
+Entities like Post have fields that only be rendered on the server, like title, excerpt,
+and content. The REST API exposes both the raw markup and the rendered version of those fields.
+For example, in the block editor, content.rendered could used as a visual preview, and
+content.raw could be used to populate the code editor.
+
+When updating these rendered fields, JavaScript is not be able to properly render arbitrary block
+markup. Therefore, it stores only the raw markup without the rendered part. And since that's a string,
+the entire field becomes a string.
+
+```ts
+type Post< C extends Context > {
+  title: RenderedText< C >;
+}
+const post = {} as Post;
+// post.title is an object with raw and rendered properties
+
+const updatablePost = {} as Updatable< Post >;
+// updatablePost.title is a string
+```
+
+### RenderedText
+
+A string that the server renders which often involves modifications from the raw source string.
+
+For example, block HTML with the comment delimiters exists in `post_content` but those comments are stripped out when rendering to a page view. Similarly, plugins might modify content or replace shortcodes.
+
+## Extending
+
+You can extend the entity record definitions using [TypeScript's declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).
+
+For example, if you're building a plugin that displays a number of views of each comment, you can add a new `numberOfViews` field to the `Comment` type like this:
+
+```ts
+// In core-data
+export interface ExtensibleComment< C extends Context > {
+	id: number;
+	// ...
+}
+
+export type Comment< C extends Context > = OmitNevers<
+	ExtensibleComment< C >
+>;
+
+// In the plugin
+import { ExtensibleComment, Comment } from '@wordpress/core-data';
+interface ExtensibleComment < C extends Context > {
+	numberOfViews: number;
+}
+
+const c : Comment = ...
```

Comment:
I tried mirroring this in some scratch code and it didn't work. TypeScript complained instead that I had an unused import and an unused interface. did we verify this works?

Reply:
Ah sorry, it didn't work – I left it there to follow up later. I added the missing `	"types": "src/types/index.ts",` entry to package.json, converged to using a namespace, and updated this example to one that does what it says.